contrib/envoyproxy/go-control-plane/cmd/serviceextensions/main.go

```diff
@@ -0,0 +1,154 @@
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/reflection"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+// AppsecCalloutExtensionService defines the struct that follows the ExternalProcessorServer interface.
+type AppsecCalloutExtensionService struct {
+	extproc.ExternalProcessorServer
+}
+
+type serviceExtensionConfig struct {
+	extensionPort   string
+	extensionHost   string
+	healthcheckPort string
+}
+
+func loadConfig() serviceExtensionConfig {
+	extensionPortInt := internal.IntEnv("DD_SERVICE_EXTENSION_PORT", 443)
+	if extensionPortInt < 1 || extensionPortInt > 65535 {
+		log.Error("service_extension: invalid port number: %d\n", extensionPortInt)
+		os.Exit(1)
+	}
+
+	healthcheckPortInt := internal.IntEnv("DD_SERVICE_EXTENSION_HEALTHCHECK_PORT", 80)
+	if healthcheckPortInt < 1 || healthcheckPortInt > 65535 {
+		log.Error("service_extension: invalid port number: %d\n", healthcheckPortInt)
+		os.Exit(1)
+	}
+
+	extensionHost := internal.IpEnv("DD_SERVICE_EXTENSION_HOST", "0.0.0.0")
+	extensionPortStr := strconv.FormatInt(int64(extensionPortInt), 10)
+	healthcheckPortStr := strconv.FormatInt(int64(healthcheckPortInt), 10)
+
+	// check if the ports are free
+	l, err := net.Listen("tcp", extensionHost+":"+extensionPortStr)
+	if err != nil {
+		log.Error("service_extension: failed to listen on extension %s:%s: %v\n", extensionHost, extensionPortStr, err)
+		os.Exit(1)
+	}
+	err = l.Close()
+	if err != nil {
+		log.Error("service_extension: failed to close listener on %s:%s: %v\n", extensionHost, extensionPortStr, err)
+		os.Exit(1)
+	}
+
+	l, err = net.Listen("tcp", extensionHost+":"+healthcheckPortStr)
+	if err != nil {
+		log.Error("service_extension: failed to listen on health check %s:%s: %v\n", extensionHost, healthcheckPortStr, err)
+		os.Exit(1)
+	}
+	err = l.Close()
+	if err != nil {
+		log.Error("service_extension: failed to close listener on %s:%s: %v\n", extensionHost, healthcheckPortStr, err)
+		os.Exit(1)
+	}
+
+	return serviceExtensionConfig{
+		extensionPort:   extensionPortStr,
+		extensionHost:   extensionHost,
+		healthcheckPort: healthcheckPortStr,
+	}
+}
+
+func main() {
+	var extensionService AppsecCalloutExtensionService
+
+	// Set the DD_VERSION to the current tracer version if not set
+	if os.Getenv("DD_VERSION") == "" {
+		if err := os.Setenv("DD_VERSION", version.Tag); err != nil {
+			log.Error("service_extension: failed to set DD_VERSION environment variable: %v\n", err)
+		}
+	}
+
+	config := loadConfig()
+
+	tracer.Start(tracer.WithAppSecEnabled(true))
+	// TODO: Enable ASM standalone mode when it is developed (should be done for Q4 2024)
+
+	go StartGPRCSsl(&extensionService, config)
+	log.Info("service_extension: callout gRPC server started on %s:%s\n", config.extensionHost, config.extensionPort)
+
+	go startHealthCheck(config)
+	log.Info("service_extension: health check server started on %s:%s\n", config.extensionHost, config.healthcheckPort)
+
+	select {}
```

Comment:
I think there are a few potential improvements here:
1. We should probably listen to SIGTERM and run cleanup / exit the program (you can use this [example](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/run/sigterm-handler/main.go) here for reference)
2. If one of the goroutines fails and exits, there is a risk that it fails silently and the program keeps running without any actual server runnning. You can solve this issue by using the `[errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup` package ([example](https://medium.com/@ninucium/golang-concurrency-patterns-for-select-done-errgroup-and-worker-pool-645bec0bd3c9) here), which will create a new context and cancel it if any of the goroutines exits with an error.

If doing 2, you can remove the free port checks from `loadConfig`, as you will be doing this anyway when you attempt to start the servers.

Please let me know if this makes sense 🙏

Reply:
Okay so, I added in 22f3e0e:
- catch signals to shutdown gracefully
- errgroup to wait for the servers to start (and check for errors)